File: javascript/packages/linter/src/rules/html-no-underscores-in-attribute-names.ts
import { ParserRule } from "../types.js"
import {
  AttributeVisitorMixin,
  StaticAttributeStaticValueParams,
  StaticAttributeDynamicValueParams,
  DynamicAttributeStaticValueParams,
  DynamicAttributeDynamicValueParams
} from "./rule-utils.js"
import { getStaticContentFromNodes } from "@herb-tools/core"
import { IdentityPrinter } from "@herb-tools/printer"
import type { LintContext, LintOffense } from "../types.js"
import type { ParseResult, HTMLAttributeNode } from "@herb-tools/core"

class HTMLNoUnderscoresInAttributeNamesVisitor extends AttributeVisitorMixin {
  protected checkStaticAttributeStaticValue({ attributeName, attributeNode }: StaticAttributeStaticValueParams): void {
    this.check(attributeName, attributeNode)
  }

  protected checkStaticAttributeDynamicValue({ attributeName, attributeNode }: StaticAttributeDynamicValueParams): void {
    this.check(attributeName, attributeNode)
  }

  protected checkDynamicAttributeStaticValue({ nameNodes, attributeNode }: DynamicAttributeStaticValueParams) {
    const attributeName = getStaticContentFromNodes(nameNodes)

    this.check(attributeName, attributeNode)
  }

  protected checkDynamicAttributeDynamicValue({ nameNodes, attributeNode }: DynamicAttributeDynamicValueParams) {
    const attributeName = getStaticContentFromNodes(nameNodes)

    this.check(attributeName, attributeNode)
  }

  private check(attributeName: string | null, attributeNode: HTMLAttributeNode): void {
    if (!attributeName) return

    if (attributeName.includes("_")) {
      this.addOffense(
        `Attribute \`${IdentityPrinter.print(attributeNode.name)}\` should not contain underscores. Use hyphens (-) instead.`,
        attributeNode.value!.location,
        "warning"
      )
    }
  }
}

export class HTMLNoUnderscoresInAttributeNamesRule extends ParserRule {
  name = "html-no-underscores-in-attribute-names"

  check(result: ParseResult, context?: Partial<LintContext>): LintOffense[] {
    const visitor = new HTMLNoUnderscoresInAttributeNamesVisitor(this.name, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_no_underscores_in_attribute_names.rs
use crate::utils::tag_utils::{get_attribute_name_literal_content, get_attributes, print_attribute_name};

use herb::nodes::HTMLOpenTagNode;
use herb::Visitor;

rule_visitor!(NoUnderscoresInAttributeNamesVisitor);
define_parser_rule!(
  HTMLNoUnderscoresInAttributeNamesRule,
  "html-no-underscores-in-attribute-names",
  Warning,
  NoUnderscoresInAttributeNamesVisitor
);

impl Visitor for NoUnderscoresInAttributeNamesVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    for attribute in get_attributes(node) {
      let checkable_name = match get_attribute_name_literal_content(attribute) {
        Some(name) => name,
        None => continue,
      };

      if checkable_name.contains('_') {
        let printed_name = print_attribute_name(attribute);
        let location = attribute
          .name
          .as_ref()
          .map(|name_node| name_node.location.clone())
          .unwrap_or_else(|| attribute.location.clone());

        self.add_offense(
          format!("Attribute `{}` should not contain underscores. Use hyphens (-) instead.", printed_name),
          location,
        );
      }
    }

    self.walk_html_open_tag_node(node);
  }
}
